### ISSUE TITLE
GridSearchCV and getting train / val loss

### ISSUE THREAD

**Issue by asmyoo** [skorch-dev/skorch#935]
Hello,

I am a bit confused about how to get the train_loss and valid_loss from history when using sklearn's GridSearchCV on NeuralNetClassifier

I read https://github.com/skorch-dev/skorch/issues/740 but I am semi confused - how should I write the callback function (seemingly called LogHistory in the link) so that I can get all the train_loss and valid_loss per each cv fold within the net.history object? Is there a way to get the train_loss and valid_loss for each fold?

**Comment by BenjaminBossan** [comment]
Let me give you a more complete example based on that thread:

```python
import numpy as np
import torch
import torch.nn.functional as F
from sklearn.datasets import make_classification
from sklearn.model_selection import GridSearchCV
from skorch import NeuralNetClassifier
from skorch.callbacks import Callback
from torch import nn

X, y = make_classification(1000, 20, n_informative=10, random_state=0)
X, y = X.astype(np.float32), y.astype(np.int64)

class ClassifierModule(nn.Module):
    def __init__(
            self,
            num_units=10,
            nonlin=F.relu,
            dropout=0.5,
    ):
        super(ClassifierModule, self).__init__()
        self.num_units = num_units
        self.nonlin = nonlin
        self.dropout = dropout

        self.dense0 = nn.Linear(20, num_units)
        self.nonlin = nonlin
        self.dropout = nn.Dropout(dropout)
        self.dense1 = nn.Linear(num_units, 10)
        self.output = nn.Linear(10, 2)

    def forward(self, X, **kwargs):
        X = self.nonlin(self.dense0(X))
        X = self.dropout(X)
        X = F.relu(self.dense1(X))
        X = F.softmax(self.output(X), dim=-1)
        return X

histories = []

class LogHistory(Callback):
    def on_train_end(self, net, *args, **kwargs):
        params = net.get_params()  # store corresponding params
        history = net.history[:]   # creates a list copy of the history
        histories.append((params, history))

net = NeuralNetClassifier(
    ClassifierModule,
    verbose=0,               # <= set verbosity to 0 to avoid too many print outputs
    train_split=False,       # <= unless really needed, turn off the skorch train/valid split, because grid search already splits data
    callbacks=[LogHistory],  # <= add the callback
)

params = {
    'lr': [0.05, 0.1],
    'module__dropout': [0, 0.5],
}

gs = GridSearchCV(net, params, refit=False, cv=3, scoring='accuracy')

print(len(histories))  # prints 12: 2x2 params and 3 CV splits
print(histories[0][0])  # prints the parameters used in the first fold
print(histories[0][1])  # prints the training history of the net from the first fold
```

Let me know if it helps or if you have more questions.

**Comment by asmyoo** [comment]
I'm confused - I thought to use the cv splits in `GridSearchCV(net, params, refit=False, cv=3, scoring='accuracy')` I had to set `train_split=False`. Also, does this mean that there is no way to write a self-contained LogHistory class (that doesn't modify `histories=[]` outside of the class)?

Thank you for all your help with this, I really appreciate it

**Comment by BenjaminBossan** [comment]
> I'm confused - I thought to use the cv splits in `GridSearchCV(net, params, refit=False, cv=3, scoring='accuracy')` I had to set `train_split=False`

The way it works is like this: You have your input data `X` and `y`. When you call `gs.fit(X, y)`, the `GridSearchCV` will split the data (in the example above 2/3 for training and 1/3 for validation). Let's call the training data part `X_gs, y_gs`. Internally, the grid search will fit the net with this data: `net.fit(X_gs, y_gs)`. Now, inside the skorch net, the data will be split _once more_ (by default 80%/20%). So from the original `X` and `y`, we first lose 1/3 to grid search and then another 1/5 to skorch's split.

In order to prevent skorch from splitting, you have to set `train_split=False`. Then skorch will use 100% of `X_gs, y_gs` for training. In most circumstances, this is what you want. You don't need skorch to set aside validation data because the grid search is already doing that for you, no need to do it twice. Does this help with the confusion?

> Also, does this mean that there is no way to write a self-contained LogHistory class (that doesn't modify `histories=[]` outside of the class)?

Do I understand correctly that you would like to have a solution that doesn't use a side effect like accessing a global variable? That's not really possible. The reason is that the grid search creates copies of the `net` and then _throws them away_. When the `gs.fit` is finished, it is impossible to recover those individual net copies. Therefore, the only way to get the histories of the nets is by using some sort of side effect. You could also write the histories to a file or save the nets at the end of the training, but you will always need some kind of side effect.

**Comment by asmyoo** [comment]
Thank you so much for the clarification! Yes what you said makes a lot of sense

Does this mean when I set `train_split=False`, I will not have valid_loss in my history since inside skorch net there is no split for validation?

Thank you again, I really appreciate all your help

**Comment by BenjaminBossan** [comment]
> Does this mean when I set `train_split=False`, I will not have valid_loss in my history since inside skorch net there is no split for validation?

Exactly. Most of the time, you don't need it, because you can use the validation results from grid search (use `gs.cv_results_` to see them). However, it depends on your use case, sometimes you may want to know the validation score history too.

**Comment by asmyoo** [comment]
Hello again,

Sorry to bother you about this again - I wrote a code snippet and was wondering if it would have the same effect as LogHistory except without creating a global variable

At this stage, I have finished performing grid search and now want the train/validation loss for each fold for only the best model (which right now is a pipeline object cls)

`
splits = list(
        StratifiedKFold(
            n_splits=grid_search_n_splits,
            random_state=random_state,
            shuffle=True,
        ).split(X_arr_train, y_arr_train)
    )

for i, split in enumerate(splits):
        # X_train, y_train = X[split[0]], y[split[0]]
        # X_val, y_val = X[split[1]], y[split[1]]
        valid_splits = skorch.dataset.ValidSplit(cv=[split])
        clf.named_steps['cls'].net.train_split = valid_splits

        clf.fit(X, y)
        history = clf.named_steps['cls'].net.history
        curves[i] = {
            'train_loss': history[:, "batches", :, 'train_loss'],
            'valid_loss': history[:, "batches", :, 'valid_loss'],
        }
`

**Comment by BenjaminBossan** [comment]
No worries, you're always welcome to ask.

I haven't run your code, so I may be missing something, but what I understand is that you try to "replay" the grid search (using only the best model) to get a record of the training history. IMO, this is not recommendable, because it's difficult to ensure that you can reproduce the steps 100%. Any small difference, e.g. a random seed change, will lead to different results. Therefore, if I was interested in the training history _exactly as encountered during the grid search_ I would go with my previous suggestion.

In general, I feel like something is still missing from my understanding.

First, why do you want to avoid the global variable? I know that it's often said it's a bad idea, but as with all things programming, context matters, and here I think it's a good solution.

Second, why do you need the history of the train run? In case you don't need the exact same history, you could also try [`cross_validate`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html), which has an option to _return the estimators_. Here some untested code:

```python
# first normal grid search
clf = Pipeline(...)
cv = ...
scoring = ...
gs = GridSearchCV(clf, cv=cv, scoring=scoring, ...)
gs.fit(X, y)
best_params = gs.best_params_

# now cross-validate
clf.set_params(**best_params)
scores = cross_validate(clf, X, y, cv=cv, scoring=scoring, return_estimator=True, ...)
for estimator in scores['estimator']:
    history = clf.named_steps['cls'].net.history
    ...
```

I'm not sure if this will 100% reproduce the result from grid search, probably not, but maybe that's not needed in your case.

**Comment by asmyoo** [comment]
Hi,

Thanks for the thorough response! Yes you are correct in interpreting my code as trying to "replay" the grid search using only the best model

Basically, I have some code already for parsing my grid search results, but adding the list of histories seems like it would be difficult for me to double-check/test that I am matching the histories results to the correct model/parameters I ran. Since I only need the best model's history, I was hoping I could simply recalculate them. Also, for code cleanliness I do not like using global variables, but I agree it is a good solution in this case. Also, I want to plot the training and validation loss for each fold, but with `train_split = False` it seems like the histories object would not have a validation loss calculated, and I'm unsure how to go about calculating it without replaying the grid search more or less

Thanks again for all your help, I really appreciate it

**Comment by BenjaminBossan** [comment]
> Yes you are correct in interpreting my code as trying to "replay" the grid search using only the best model

Okay. So if you want to ensure that you get 100% the same results, I don't see any other way than to rerun the grid search and save a copy of the histories, otherwise the results will probably be similar but not identical. If 100% reproduction is not necessary, something like the code you suggested or using `cross_validate` with `return_estimator=True` should work.

> adding the list of histories seems like it would be difficult for me to double-check/test that I am matching the histories results to the correct model/parameters I ran

This is why I suggested to add the model parameters to the list, that way you can match each training history to its corresponding parameters.

```python
class LogHistory(Callback):
    def on_train_end(self, net, *args, **kwargs):
        params = net.get_params()  # store corresponding params
        history = net.history[:]   # creates a list copy of the history
        histories.append((params, history))
```

> Also, I want to plot the training and validation loss for each fold, but with train_split = False it seems like the histories object would not have a validation loss calculated, and I'm unsure how to go about calculating it without replaying the grid search more or less

If you really need the validation loss history, it's fine to enable the train split, but as I mentioned, it means less data is used for training. If you're only interested in the _final_ validation loss (i.e. once the model stopped training), the grid search `cv_results_` should contain those. (But note that since `GridSearchCV` uses a different split than skorch, the results are not 100% the same.)

**Comment by asmyoo** [comment]
Thank you again so much for your help, I appreciate it! Going to use your LogHistory method for my code, thanks again for explaining this all in detail
